apps/src/music/progress/ConditionsChecker.ts

```diff
@@ -17,21 +17,34 @@ export default class ConditionsChecker {
   }
 
   // Accumulate a satisfied condition.
-  addSatisfiedCondition(id: string) {
-    this.currentSatisfiedConditions.push(id);
+  addSatisfiedCondition(condition: KnownCondition) {
+    if (!this.isConditionSatisfied(condition)) {
+      this.currentSatisfiedConditions.push(condition);
+    }
+  }
+
+  private isConditionSatisfied(condition: KnownCondition) {
+    return this.currentSatisfiedConditions.some(
+      currentSatisfiedCondition =>
+        JSON.stringify(currentSatisfiedCondition) === JSON.stringify(condition)
```

Comment:
We really shouldn't use JSON.stringify() for comparison's like this - it's relatively slow and expensive (potentially very with large json) and the key order isn't guaranteed which can cause false negatives.

Can we just drill into a few specific props and compare those? Otherwise, something like lodash's `isEqual` is a better bet.

Reply:
Fair points, though just curious about the state of this:

> the key order isn't guaranteed which can cause false negatives

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify has been [updated](https://github.com/mdn/content/issues/16583) to say: 

> Properties are visited using the same algorithm as [Object.keys()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys), which has a well-defined order and is stable across implementations

Reply:
As an alternate approach - could we use a `Set` of stringified/serialized `Condition`s so we don't need to manually check if it already exists when adding?

Reply:
That mozdev note only says that key order will be consistent on the same object. It's not claiming for consistency across objects. e.g.,

```
console.log(JSON.stringify({a : 1, b : 2}));  // outputs {"a":1,"b":2}
console.log(JSON.stringify({b : 2, a : 1}));  // outputs {"b":2,"a":1}
```

Those two objects are definitely identical, but would fail a stringified json check.

Reply:
> Those two objects are definitely identical, but would fail a stringified json check.

Ah, of course.  I've made the switch to `_.isEqual`.

> As an alternate approach - could we use a Set of stringified/serialized Conditions so we don't need to manually check if it already exists when adding?

I tried using a `Set` but without native object comparisons, which haven't arrived [yet](https://stackoverflow.com/a/29759699), it seems like it would suffer from the same `JSON.stringify` limitations, and we'd have to `JSON.parse` quite often to use what's in it.